tests/utils/test_configdict.py

```diff
@@ -0,0 +1,129 @@
+import pytest
+
+from impactlab_tools.utils.configdict import ConfigDict, gather_configtree
+
+
+@pytest.fixture
+def simple_nested_tree():
+    return {'a': 1, 'b': {'a': 2}, 'c': 3, 'd-4': 4, 'e_5': 5, 'F': 6}
+
+
+def test_configdict_climbs_tree(simple_nested_tree):
+    conf = ConfigDict(simple_nested_tree)
+    assert conf['c'] == 3
+
+
+def test_configdict_prefer_child(simple_nested_tree):
+    conf = gather_configtree(simple_nested_tree)['b']
+    assert conf['a'] == 2
+
+
+def test_configdict_selfreference(simple_nested_tree):
+    conf = gather_configtree(simple_nested_tree)
+    conf['b'] = conf
+    assert conf['b'] == conf
+
+
+def test_configdict_caseinsensitive_keys(simple_nested_tree):
+    conf = gather_configtree(simple_nested_tree)
+    assert conf['A'] == 1
+    assert conf['f'] == 6
+
+    conf['Z'] = 'foobar'
+    assert conf['z'] == 'foobar'
+
+
+def test_configdict_normalize(simple_nested_tree):
+    """Normalize underscores and hyphens in keys
+    """
+    conf = ConfigDict(simple_nested_tree)
+    assert conf['d_4'] == 4
+    assert conf['e-5'] == 5
+
+    conf['y-7'] = 'foobar'
+    assert conf['y_7'] == 'foobar'
+
+    conf['z-9'] = 'foobar'
+    assert conf['z_9'] == 'foobar'
+
+
+def test_configdict_throws_keyerror(simple_nested_tree):
+    conf = ConfigDict(simple_nested_tree)
+    with pytest.raises(KeyError):
+        conf['foobar']
+
+
+def test_configdict_merge(simple_nested_tree):
+    conf1 = ConfigDict(simple_nested_tree)
+    conf2 = ConfigDict(simple_nested_tree)
+    conf2['a'] = 9
+
+    conf_merge = conf1.merge(conf2)
+
+    assert conf_merge.data == conf2.data
+    assert conf_merge.parent == conf1.parent
+
+
+def test_configdict_merge_parentswap(simple_nested_tree):
+    conf1 = gather_configtree(simple_nested_tree)['b']
+
+    nested_tree_mod = simple_nested_tree.copy()
+    nested_tree_mod['a'] = 9
+    conf2 = gather_configtree(nested_tree_mod)['b']
+
+    conf_merge = conf1.merge(conf2, xparent=True)
+
+    assert conf_merge.data == conf2.data
+    assert conf_merge.parent == conf2.parent
+
+
+def test_configdict_key_access_stack(simple_nested_tree):
+    """Test ConfigDict adds values to ``self.key_access_stack`` on query
+    """
+    conf = ConfigDict(simple_nested_tree)
+    assert conf.key_access_stack == {}
+    conf['a']
+    assert 'a' in list(conf.key_access_stack.keys())
+
+
+def test_configdict_key_access_stack_nested(simple_nested_tree):
+    """Test ConfigDict.key_access_stack sores keys in appropriate configdict
+    """
+    conf = gather_configtree(simple_nested_tree)
+    nested = conf['b']
+    nested['a']
+    nested['f']
+
+    assert list(nested.key_access_stack.keys()) == ['a']
+
+    top_keys = list(nested.parent.key_access_stack.keys())
+    top_keys.sort()
+    assert top_keys == ['b', 'f']
```

Comment:
Rather than `nested.parent`, it's more clear to just use `conf`, and then it should also report the access of `['b']['a']` somehow.

Reply:
Got the first part in new commit.

> and then it should also report the access of ['b']['a'] somehow.

For this part I think it's important that each configdict only keeps track or its own local dictionary access. Otherwise I think things become a mess when nestings and their relationships move, merge and change... every child and parent node would need to have updated knowledge of the larger tree, etc. That's too complex, right?

*But*, see my response here https://github.com/ClimateImpactLab/impactlab-tools/pull/437#discussion_r317284214?